subt/ros/proxy/traversability.cc

```diff
@@ -0,0 +1,769 @@
+            pt.z - expected_ground < config_.max_dip_down)
+        {
+          if (pt_dist_from_sensor <= config_.visible_ground_min)
+          {
+            obscured_view = true;
+          }
+          if (cnt++ % (config_.depth_subsampling / config_.depth_image_stride / config_.depth_image_stride) == 0)
+          {
+            points.emplace_back(pt.x, pt.y, pt.z);
+          }
+        }
+      }
+      else if (horizontal.at<uint8_t>(v, u))
+      {
+        if (pt.z - robot_xyz.z() > config_.max_bump_height &&
+            pt_dist_from_sensor <= config_.visible_ground_min)
+        {
+          obscured_view = true;
+        }
+
+        if (pt_dist_from_sensor >= config_.visible_ground_min &&
+            pt_dist_from_sensor <= config_.visible_ground_max)
+        {
+          ground_visible = true;
+        }
+
+        expected_ground = pt.z;
+      }
+    }
+
+    // We react to missing ground only when the robot is looking sufficiently
+    // horizontally or downwards. Otherwise, the robot may be climbing up a
+    // ridge and does not even have a chance to see the ground.
+    if (pt_down_near.z() - robot_xyz.z() <= 0) {
+      // If
+      // a) We haven't seen any ground in this direction;
+      // b) The view is not obstructed by an obstacle;
+      // c) The robot is not pitched/rolled in a way that seeing
+      //    ground in this direction is impossible;
+      // then we should be worried.
+      if (!ground_visible && !obscured_view)
+      {
+        points.push_back(pt_down_near);
+      }
+    }
+  }
+
+  InsertObservation(std::move(points), camera_frame_id,
+                    config_.max_depth_observations);
+}
+
+float median(std::vector<float> values)
+{
+  const size_t middle = values.size() / 2;
+  std::nth_element(values.begin(), values.begin() + middle, values.end());
+  return values[middle];
+}
+
+void Traversability::OnScan(const sensor_msgs::LaserScan::ConstPtr& msg)
+{
+  std::optional<tf::StampedTransform> sensor_pose =
+    GetTransform(config_.world_frame_id, msg->header.frame_id, msg->header.stamp);
+  if (!sensor_pose) return;
+
+  std::optional<tf::StampedTransform> robot_pose =
+    GetTransform(config_.world_frame_id, config_.robot_frame_id, msg->header.stamp);
+  if (!robot_pose) return;
+  const auto& robot_xyz = robot_pose->getOrigin();
+
+  Observation points;
+  points.reserve(msg->ranges.size());
+  size_t cnt = 0;
+  tf::Transform rot;
+  tf::Vector3 beam;
+  for (size_t i = 2; i  + 2 < msg->ranges.size(); ++i)
+  {
+    const float angle = msg->angle_min + i * msg->angle_increment;
+    // Fog filtering.
+    const float range = median({
+        msg->ranges[i - 2],
+        msg->ranges[i - 1],
+        msg->ranges[i],
+        msg->ranges[i + 1],
+        msg->ranges[i + 2]});
+    if (range <= 0 || range >= config_.max_lidar_range)  // Skip invalid/missing/ignored measurements.
+    {
+      continue;
+    }
+
+    beam.setX(range);
+    rot.getRotation().setRPY(0, 0, angle);
```

Comment:
s/getRotation/getBasis
getBasis returs a reference which makes sense to modify, getRotation returns a new quaternion that this code modifies and then discards :-/

Reply:
fixed